### PR TITLE
Adding the BreakpointPartition layout algorithm

### PR DESCRIPTION
**Pull request by GreenGremlin**
I was impressed by your BSTLinearPartition algorithm. I've added a BreakpointPartition algorithm that should improve on it's efficiency. Rather than blindly perform a binary search from the maximum width. It starts from the ideal width (The ideal width being the width if the images fit perfectly in the given container.) and expands to the next width that will allow an item to move up a row. This algorithm will find the exact width that produces the "ideal" layout and should generally find it in two or three passes.

I've implemented and tested this algorithm in Python, but have not tested this version yet. If I have time, I'll add some unit tests, next week.

### CODE REVIEW

File: lib/BreakpointPartition.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    currentRowWidth: 0,
+    rowCount: 1,
+    // the largest possible step to the next breakpoint is the smallest image ratio
+    nextBreakpoint: Math.min.apply(null, imageRatioSequence)
+  }
+  const finalLayout = imageRatioSequence.reduce((layout, itemWidth) => {
+    const rowWidth = layout.currentRowWidth + itemWidth;
+    let nextBreakpoint;
+    if (rowWidth > expectedWidth) {
+      nextBreakpointForCurrentRow = layout.currentRowWidth - expectedWidth;
```
Comment (GreenGremlin) [inline lib/BreakpointPartition.js]: Here is the key difference in this algorithm. When an image is too wide to fit on the current row, I'm calculating how much more width would be needed to include it. If it requires less space than any previous row needed, I track it. When we've passed through all rows, if the current layout produces too many rows, I make another pass at the smallest width that will allow an image to move up a row.

**Comment by GreenGremlin** [comment]
I tried forking your pen, but it's having trouble finding perfectLayout.

http://codepen.io/xgecko/pen/pjBWEE

**Comment by GreenGremlin** [comment]
Still working on this.

**Comment by GreenGremlin** [comment]
Ok, it's tested and working.

http://codepen.io/xgecko/full/pjBWEE/

**Comment by GreenGremlin** [comment]
I created a JSPerf to see if the breakpoint partitioning was actually faster. I find it interesting that in Chrome the two are about the same, but in Firefox breakpoint partitioning is more than twice as fast.

http://jsperf.com/bst-vs-breakpoint-partitioning

**Comment by axyz** [comment]
Hey, very nice job and idea. 
I'm going to give a deeper look at it and will try to merge it on the week end.
I'm using chrome and I got it 48% faster with breakpoint partitioning on your jsperf.

**Comment by GreenGremlin** [comment]
Cool, thanks!

Check this out! [Pure css perfect layout](http://codepen.io/xgecko/pen/EVdvNd)

**Pull request merged**